File: tests/docs/python/test_attribute_read.py
# Copyright (c) 2025-2026, NVIDIA CORPORATION. All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto.  Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.

"""Tests for the stage attribute read API (``Renderer.read_attribute`` /
``read_array_attribute`` and their async variants).

Reads target schema-known attributes the runtime exposes to Fabric:
``omni:rtx:rtpt:maxBounces`` on the RenderProduct (a 32-bit integer; the
runtime may report it as either int32 or uint32) and ``points`` on
``/World/Plane`` (a float3 array authored in ``ovrtx-test-base-geometry.usda``).
"""

from pathlib import Path

import numpy as np
import ovrtx

TEST_BASE_PATH = str((Path(__file__).parent / "../data/ovrtx-test-base.usda").resolve())


def _load_base(renderer):
    renderer.open_usd(TEST_BASE_PATH)
    renderer.reset()


def test_read_scalar_attribute_cpu(renderer):
    """Write a schema-known int32 scalar and read it back."""
    _load_base(renderer)

    # Seed a known value so the read has something deterministic to observe.
    renderer.write_attribute(
        prim_paths=["/Render/Camera"],
        attribute_name="omni:rtx:rtpt:maxBounces",
        tensor=np.array([17], dtype=np.int32),
    )

    # [snippet:doc-read-attribute-scalar]
    # Read a scalar attribute — one value per prim. The returned ManagedDLTensor
    # is DLPack-compatible; np.from_dlpack() gives a zero-copy numpy view.
    tensor = renderer.read_attribute(
        attribute_name="omni:rtx:rtpt:maxBounces",
        prim_paths=["/Render/Camera"],
    )
    values = np.from_dlpack(tensor)
    # [/snippet:doc-read-attribute-scalar]

    assert isinstance(tensor, ovrtx.ManagedDLTensor)
    # maxBounces is a 32-bit integer; the runtime may report it as int32 or
    # uint32 depending on the schema — compare the value, not the sign-ness.
    assert values.dtype.itemsize == 4 and values.dtype.kind in ("i", "u")
    assert values.shape == (1,)
    assert int(values[0]) == 17


def test_read_scalar_attribute_into_dest(renderer):
    """Pass a pre-allocated tensor as ``dest``; verify data lands in it."""
    _load_base(renderer)

    renderer.write_attribute(
        prim_paths=["/Render/Camera"],
        attribute_name="omni:rtx:rtpt:maxBounces",
        tensor=np.array([11], dtype=np.int32),
    )

    # [snippet:doc-read-attribute-dest-tensor]
    # Pre-allocate the destination. The read writes directly into `dest`; the
    # returned tensor is a handle to the same memory — both aliases are valid.
    # The dtype must match how the runtime stores the attribute.
    dest = np.empty((1,), dtype=np.uint32)
    renderer.read_attribute(
        attribute_name="omni:rtx:rtpt:maxBounces",
        prim_paths=["/Render/Camera"],
        dest=dest,
    )
    # `dest` now holds the attribute value.
    # [/snippet:doc-read-attribute-dest-tensor]

    assert int(dest[0]) == 11


def test_read_array_attribute(renderer):
    """Read ``points`` (``float3[]``) from the Plane mesh."""
    _load_base(renderer)

    # [snippet:doc-read-array-attribute]
    # Arrays are returned as dict[prim_path, ManagedDLTensor]. Iteration order
    # matches the input prim_paths so you can zip() against the request.
    tensors = renderer.read_array_attribute(
        attribute_name="points",
        prim_paths=["/World/Plane"],
    )
    for path, tensor in tensors.items():
        values = np.from_dlpack(tensor)
        print(f"{path}: {values.size} elements, dtype={values.dtype}")
    # [/snippet:doc-read-array-attribute]

    assert list(tensors.keys()) == ["/World/Plane"]
    values = np.from_dlpack(tensors["/World/Plane"])
    # ovrtx-test-base-geometry.usda authors 4 float3 points on the Plane.
    assert values.dtype == np.float32
    assert values.size == 4 * 3


def test_read_attribute_async(renderer):
    """Exercise the two-phase async lifecycle for an attribute read."""
    _load_base(renderer)

    renderer.write_attribute(
        prim_paths=["/Render/Camera"],
        attribute_name="omni:rtx:rtpt:maxBounces",
        tensor=np.array([5], dtype=np.int32),
    )

    # [snippet:doc-read-attribute-async]
    # Two-phase async: wait() → PendingFetch, then fetch() → ManagedDLTensor.
    op = renderer.read_attribute_async(
        attribute_name="omni:rtx:rtpt:maxBounces",
        prim_paths=["/Render/Camera"],
    )
    pending = op.wait()
    tensor = pending.fetch()
    values = np.from_dlpack(tensor)
    # [/snippet:doc-read-attribute-async]

    assert isinstance(op, ovrtx.Operation)
    assert isinstance(pending, ovrtx.PendingFetch)
    assert isinstance(tensor, ovrtx.ManagedDLTensor)
    assert int(values[0]) == 5


def test_read_attribute_cuda_dest(renderer):
    """Read directly into a GPU (CUDA) destination tensor via DLPack."""
    import warp as wp  # via warp-lang; provides DLPack-compatible CUDA arrays

    _load_base(renderer)

    renderer.write_attribute(
        prim_paths=["/Render/Camera"],
        attribute_name="omni:rtx:rtpt:maxBounces",
        tensor=np.array([9], dtype=np.int32),
    )

    # [snippet:doc-read-attribute-cuda-dest]
    # Allocate a CUDA destination through Warp (any DLPack-compatible CUDA
    # allocator works). The read writes directly into GPU memory; pass a CUDA
    # stream handle so the read is ordered on the caller's stream.
    dest = wp.empty(1, dtype=wp.uint32, device="cuda:0")
    stream = wp.Stream(device=dest.device)
    renderer.read_attribute(
        attribute_name="omni:rtx:rtpt:maxBounces",
        prim_paths=["/Render/Camera"],
        dest=dest,
        cuda_stream=stream.cuda_stream,
    )
    wp.synchronize_stream(stream)
    # [/snippet:doc-read-attribute-cuda-dest]

    # Copy back to host and verify the value.
    host = dest.numpy()
    assert int(host[0]) == 9

File: tests/docs/c/test_attribute_read.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
// property and proprietary rights in and to this material, related
// documentation and any modifications thereto. Any use, reproduction,
// disclosure or distribution of this material and related documentation
// without an express license agreement from NVIDIA CORPORATION or
// its affiliates is strictly prohibited.

// Tests for ovrtx_read_attribute / ovrtx_fetch_read_result / ovrtx_release_read_result.
// Scalar test: write-then-read omni:rtx:rtpt:maxBounces on /Render/Camera.
// Array test: read the authored "points" attribute from /World/Plane.

#include <gtest/gtest.h>
#include "helpers.h"

#include <ovrtx/ovrtx_attributes.h>

#include <cstdint>
#include <cstring>
#include <string>
#include <vector>

class AttributeReadTest : public ::testing::Test {
protected:
    static void SetUpTestSuite() {
        TestConfig tc("AttributeReadTest");
        ovrtx_result_t result = ovrtx_create_renderer(&tc.config, &renderer_);
        ASSERT_API_SUCCESS(result.status);
    }

    static void TearDownTestSuite() {
        if (renderer_) {
            ovrtx_destroy_renderer(renderer_);
            renderer_ = nullptr;
        }
    }

    static void load_base() {
        ovrtx_op_wait_result_t wait_result;
        ovrtx_enqueue_result_t eq;

        eq = ovrtx_reset_stage(renderer_);
        ASSERT_API_SUCCESS(eq.status);
        ASSERT_API_SUCCESS(ovrtx_wait_op(renderer_, eq.op_index, ovrtx_timeout_infinite, &wait_result).status);
        ASSERT_NO_OP_ERRORS(wait_result);

        std::string scene = get_docs_test_data_dir() + "/ovrtx-test-base.usda";
        eq = ovrtx_open_usd_from_file(renderer_, {scene.c_str(), scene.size()});
        ASSERT_API_SUCCESS(eq.status);
        ASSERT_API_SUCCESS(ovrtx_wait_op(renderer_, eq.op_index, ovrtx_timeout_infinite, &wait_result).status);
        ASSERT_NO_OP_ERRORS(wait_result);

        eq = ovrtx_reset(renderer_, 0.0);
        ASSERT_API_SUCCESS(eq.status);
        ASSERT_API_SUCCESS(ovrtx_wait_op(renderer_, eq.op_index, ovrtx_timeout_infinite, &wait_result).status);
        ASSERT_NO_OP_ERRORS(wait_result);
    }

    // Write a single int32 value to the given prim/attribute so the subsequent
    // read has a deterministic value to observe.
    static void write_int32(char const* prim, char const* attribute, int32_t value) {
        ovx_string_t p = ovx_str(prim);
        size_t count = 1;
        DLDataType int32_type = {kDLInt, 32, 1};
        DLTensor tensor = ovrtx_make_write_cpu_tensor(&value, &count, int32_type);

        ovrtx_input_buffer_t buffer{};
        buffer.tensors = &tensor;
        buffer.tensor_count = 1;

        ovrtx_binding_desc_or_handle_t binding = ovrtx_make_binding_desc(
            &p, 1, ovx_str(attribute), OVRTX_SEMANTIC_NONE, int32_type);

        ovrtx_enqueue_result_t eq =
            ovrtx_write_attribute(renderer_, &binding, &buffer, OVRTX_DATA_ACCESS_SYNC);
        ASSERT_API_SUCCESS(eq.status);
        ovrtx_op_wait_result_t wait_result;
        ASSERT_API_SUCCESS(ovrtx_wait_op(renderer_, eq.op_index, ovrtx_timeout_infinite, &wait_result).status);
        ASSERT_NO_OP_ERRORS(wait_result);
    }

    static ovrtx_renderer_t* renderer_;
};

ovrtx_renderer_t* AttributeReadTest::renderer_ = nullptr;

TEST_F(AttributeReadTest, ReadScalarAttribute) {
    load_base();
    write_int32("/Render/Camera", "omni:rtx:rtpt:maxBounces", 17);

    // [snippet:doc-read-attribute-scalar-c]
    // Describe the attribute to read: one prim, name, element type matching
    // how the runtime stores it (maxBounces is a 32-bit unsigned integer).
    ovx_string_t rp = ovx_str("/Render/Camera");
    DLDataType uint32_type = {kDLUInt, 32, 1};
    ovrtx_binding_desc_or_handle_t binding = ovrtx_make_binding_desc(
        &rp, 1, ovx_str("omni:rtx:rtpt:maxBounces"), OVRTX_SEMANTIC_NONE, uint32_type);

    // Enqueue the read. Pass NULL for read_dest so ovrtx allocates internal storage.
    ovrtx_read_handle_t read_handle = 0;
    ovrtx_enqueue_result_t eq = ovrtx_read_attribute(renderer_, &binding, nullptr, &read_handle);
    ASSERT_API_SUCCESS(eq.status);

    ovrtx_op_wait_result_t wait_result;
    ASSERT_API_SUCCESS(ovrtx_wait_op(renderer_, eq.op_index, ovrtx_timeout_infinite, &wait_result).status);
    ASSERT_NO_OP_ERRORS(wait_result);

    // Fetch the DLPack tensor(s). Scalar reads produce buffer_count == 1 with
    // shape [prim_count].
    ovrtx_read_output_t output{};
    ASSERT_API_SUCCESS(ovrtx_fetch_read_result(renderer_, read_handle, ovrtx_timeout_infinite, &output).status);
    ASSERT_EQ(output.buffer_count, 1u);
    ASSERT_EQ(output.prim_count, 1u);

    DLTensor const& t = output.buffers[0].dl;
    uint32_t value = *static_cast<uint32_t const*>(t.data);

    // Release the read result when done. The output pointers are invalidated.
    ovrtx_cuda_sync_t no_sync{};
    ASSERT_API_SUCCESS(ovrtx_release_read_result(renderer_, output.map_handle, no_sync).status);
    // [/snippet:doc-read-attribute-scalar-c]

    EXPECT_EQ(value, 17u);
}

TEST_F(AttributeReadTest, ReadArrayAttribute) {
    load_base();

    // [snippet:doc-read-array-attribute-c]
    // Array attributes are variable-length per prim. The dtype in the binding
    // is (code, bits, lanes) — lanes expresses multi-component element types.
    // `points` is float3[], so request float32 with lanes=3: one element per
    // point, three lanes per point. Override is_array=true (the helper default
    // is false).
    ovx_string_t prim = ovx_str("/World/Plane");
    DLDataType point3f_type = {kDLFloat, 32, 3};
    ovrtx_binding_desc_or_handle_t binding = ovrtx_make_binding_desc(
        &prim, 1, ovx_str("points"), OVRTX_SEMANTIC_NONE, point3f_type);
    binding.binding_desc.attribute_type.is_array = true;

    ovrtx_read_handle_t read_handle = 0;
    ovrtx_enqueue_result_t eq = ovrtx_read_attribute(renderer_, &binding, nullptr, &read_handle);
    ASSERT_API_SUCCESS(eq.status);

    ovrtx_op_wait_result_t wait_result;
    ASSERT_API_SUCCESS(ovrtx_wait_op(renderer_, eq.op_index, ovrtx_timeout_infinite, &wait_result).status);
    ASSERT_NO_OP_ERRORS(wait_result);

    // For arrays the output has one tensor per prim (buffer_count == prim_count).
    ovrtx_read_output_t output{};
    ASSERT_API_SUCCESS(ovrtx_fetch_read_result(renderer_, read_handle, ovrtx_timeout_infinite, &output).status);
    ASSERT_TRUE(output.is_array);
    ASSERT_EQ(output.prim_count, 1u);
    ASSERT_EQ(output.buffer_count, 1u);

    DLTensor const& t = output.buffers[0].dl;
    // The Plane has 4 float3 points. The C API returns lane-based attribute
    // tensors, so this is shape=[4], dtype={kDLFloat, 32, 3}.
    ASSERT_EQ(t.ndim, 1);
    ASSERT_EQ(t.shape[0], 4);
    ASSERT_EQ(t.dtype.code, kDLFloat);
    ASSERT_EQ(t.dtype.bits, 32u);
    ASSERT_EQ(t.dtype.lanes, 3u);
    int64_t element_count = t.shape[0] * t.dtype.lanes;

    ovrtx_cuda_sync_t no_sync{};
    ASSERT_API_SUCCESS(ovrtx_release_read_result(renderer_, output.map_handle, no_sync).status);
    // [/snippet:doc-read-array-attribute-c]

    EXPECT_EQ(element_count, 12);
}
